crates/matrix-sdk/src/room/mod.rs

```diff
@@ -600,6 +690,48 @@ impl Room {
             .collect())
     }
 
+    pub async fn members_with_dynamic_adapters(
+        &self,
+        page_size: usize,
+    ) -> (impl Stream<Item = Vec<VectorDiff<MemberListEntry>>>, MemberListDynamicEntriesController)
+    {
+        let list = self
+            .members(RoomMemberships::all())
+            .await
+            .unwrap()
+            .iter()
+            .map(|m| MemberListEntry::Filled(m.user_id().to_owned()))
+            .collect::<Vec<_>>();
```

Comment:
The fact you await and iter+collect before generating the stream basically breaks the purpose of this PR.

What are we trying to do here?

1. Be able to _load_ the members page-by-page, and to filter them. That's done with the fact we return a `Stream<Item = Vec<VectorDiff<…>>>` and some help from `eyeball_im_util`. We control this with `MemberListDynamicEntriesController`, that's good. But, here, we assume that all members are loaded beforehand, before returning the `Stream`. And fetfching all the members can take time, hence…
2. Be able to _fetch_ the members from the store or the network via a `Stream`. Instead of collecting all members, and generating the `Stream`, we want to collect the members inside an `ObservableVector`, then subscribe to it, and boom, we have our `Stream`.

What we need is probably a `members: ObservableVector` _field_ somewhere that contains all the loaded members. I don't think that putting it inside `Room` is a good idea. Can we create a new type for that? We could deprecate the `Room::members` and `::members_no_sync` API after that. I think we must not _improve_ but _rewrite_ the way we manipulate members.

You're close to get something cool. The problem is that this code block above breaks 50% of the usefulness of using `Stream` here.

Reply:
We should probably have a `RoomMemberList` interface, similar to the `RoomList` one.
This `RoomMemberList` could expose 2 methods : 
-  `entries_with_dynamic_adapters`
-  `loading_state` 

Then inside this object, we should be able to be notified of members requests results and room member changes  from the sync.

Reply:
@timokoesters Why having `sdk_base::rooms::normal::Room::members()` to return an `Iterator` would simplify things? I don't get it.

---

@ganfra The problematics aren't the same than the `RoomList`. The `RoomList` receives updates sync after sync. The member list is already loaded in the store, or is updated manually, but it's not a running task in the background. Why do you want `loading_state` for example?

Reply:
To know we are currently fetching some room members?